test/e2e/metrics_test.go

```diff
@@ -0,0 +1,123 @@
+package e2e
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+	test "github.com/openshift/cluster-kube-apiserver-operator/test/library"
+)
+
+func TestMetricsRegistration(t *testing.T) {
+	kubeConfig, err := test.NewClientConfigForTest()
+	require.NoError(t, err)
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+	restConfig, err := test.NewClientConfigForTest()
+	require.NoError(t, err)
+	restConfig = setRESTConfigDefaults(restConfig)
+	restClient, err := rest.RESTClientFor(restConfig)
+	require.NoError(t, err)
+
+	// get list of operator pods
+	pods, err := kubeClient.CoreV1().Pods(operatorclient.OperatorNamespace).List(metav1.ListOptions{})
+	require.NoError(t, err)
+
+	// we just care about the one we expect
+	require.GreaterOrEqual(t, len(pods.Items), 1)
+	pod := pods.Items[0]
+
+	metrics := getMetricsForPod(t, pod, kubeConfig, restClient)
+	t.Logf("Retrieved %d metrics.", len(metrics))
+	if len(metrics) == 0 {
+		t.Fatal("No metrics retrieved.")
+	}
+}
+
+func getMetricsForPod(t *testing.T, pod v1.Pod, kubeConfig *rest.Config, restClient *rest.RESTClient) map[string]string {
+	require.Equal(t, v1.PodRunning, pod.Status.Phase)
+	transport, upgrader, err := spdy.RoundTripperFor(kubeConfig)
+	require.NoError(t, err)
+	request := restClient.Post().Resource("pods").Namespace(pod.Namespace).Name(pod.Name).SubResource("portforward")
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, request.URL())
```

Comment:
why do we need a port forwarder? This sounds super complex.

Reply:
Not saying it is not necessary (maybe it is), but it surprises me as a reader and needs a comments explaining in 2 or 3 lines why.

Reply:
I've added some comments. AFAIK, port forwarding is needed in order for test to run outside of the cluster.